src/legacy/core_plugins/expressions/public/np_ready/public/expression_renderer.tsx

```diff
@@ -17,47 +17,90 @@
  * under the License.
  */
 
-import { useRef, useEffect } from 'react';
+import { useRef, useEffect, useState } from 'react';
 import React from 'react';
-import { ExpressionAST, IExpressionLoaderParams, IInterpreterResult } from './types';
-import { IExpressionLoader, ExpressionLoader } from './loader';
+import { EuiProgress } from '@elastic/eui';
+import { ExpressionAST, IExpressionLoaderParams } from './types';
+import { ExpressionLoader } from './loader';
 
 // Accept all options of the runner as props except for the
 // dom element which is provided by the component itself
 export interface ExpressionRendererProps extends IExpressionLoaderParams {
   className: string;
   expression: string | ExpressionAST;
-  /**
-   * If an element is specified, but the response of the expression run can't be rendered
-   * because it isn't a valid response or the specified renderer isn't available,
-   * this callback is called with the given result.
-   */
-  onRenderFailure?: (result: IInterpreterResult) => void;
+  renderError?: (
+    errorType: 'data' | 'render',
+    error?: string | null
+  ) => React.ReactElement | React.ReactElement[];
+}
+
+interface State {
+  isLoading: boolean;
+  hasError: boolean;
+  errorType: 'data' | 'render' | null;
+  errorMessage: string | null;
 }
 
 export type ExpressionRenderer = React.FC<ExpressionRendererProps>;
 
-export const createRenderer = (loader: IExpressionLoader): ExpressionRenderer => ({
+const defaultState: State = {
+  isLoading: false,
+  hasError: false,
+  errorType: null,
+  errorMessage: null,
+};
+
+export const ExpressionRendererImplementation = ({
   className,
   expression,
-  onRenderFailure,
   ...options
 }: ExpressionRendererProps) => {
   const mountpoint: React.MutableRefObject<null | HTMLDivElement> = useRef(null);
   const handlerRef: React.MutableRefObject<null | ExpressionLoader> = useRef(null);
+  const [state, setState] = useState<State>(defaultState);
 
   useEffect(() => {
     if (mountpoint.current) {
+      setState({ ...defaultState, isLoading: true });
+
       if (!handlerRef.current) {
-        handlerRef.current = loader(mountpoint.current, expression, options);
+        handlerRef.current = new ExpressionLoader(mountpoint.current, expression, options);
+
+        // Only registers one subscriber
+        handlerRef.current.data$.subscribe({
+          next: () => {
+            setState(defaultState);
+          },
+          error: err => {
+            setState(prevState => ({
+              ...prevState,
+              isLoading: false,
+              hasError: true,
+              errorType: 'data',
+              errorMessage: err.toString(),
+            }));
+          },
+        });
+        handlerRef.current.render$.subscribe({
+          next: () => {
+            setState(() => ({
+              ...defaultState,
+              isLoading: false,
+            }));
+          },
+          error: err => {
```

Comment:
This is here for `renderers` to signal errors (not sure we need this). For example when i mouse over a bar and tooltip should be shown, renderer can get into error state and would use `error` handler provided to it to signal it.

- what to do with uncaught errors in renderers ? 
- what to do with uncaught errors inside expression functions ? 
- can we handle errors in a common way ? (currently it behaves differently depending on the loader you are using)
  - using expression executor will give you either { type: 'error' }  or it will throw
  - using expression loader will not throw, also it won't return any data ... but it will emit an error on $data observable
  - using expression renderer and providing it with invalid data will emit an error on $render observable
  - using expression loader which generates invalid data might emit error on $data observable or on $render observable